lib/symboldatabase.cpp

```diff
@@ -4202,8 +4204,41 @@ static const Token* skipPointersAndQualifiers(const Token* tok)
     return tok;
 }
 
+const Token* skipTypeOrFunctionIdentifier(const Token* tok)
+{
+    if (tok == nullptr)
+        return nullptr;
+    while (tok->next() != nullptr && (tok->next()->str() == "::" || tok->next()->str() == "<"))
+    {
+        if (tok->next()->str() == "::")
+            tok = tok->tokAt(2);
+        else
+            tok = tok->linkAt(1);
+        if (tok == nullptr)
+            return nullptr;
+    }
+    return tok->next();
+}
+
+//Could this token be the first token of a call to a function instead of the declaration of a variable
+bool isFunctionCallInsteadOfVariableDeclaration(const Token* tok, const Scope& scope, const SymbolDatabase& db)
+{
+    auto start = tok;
+    tok = skipTypeOrFunctionIdentifier(tok);
+    if (tok == nullptr)
+        return false; //It's probably neither but it's not a function call
+    if (tok->str() != "(") //X y isn't a function call
+        return false;
+    if (tok->astOperand2() != nullptr && tok->astOperand2()->str() == ",")// X(a, b) isn't a variable declaration
+        return true;
+    return db.findFunction(start, &scope) != nullptr;
+}
+
 bool Scope::isVariableDeclaration(const Token* const tok, const Token*& vartok, const Token*& typetok) const
 {
+    if (isFunctionCallInsteadOfVariableDeclaration(tok, *this, *this->check))
+        return false;
```

Comment:
In some unit tests using nullscope ```*this->check``` is null. This doesn't cause the tests to fail (it probably should) but does cause address sanitizer to fail. I don't think it will affect the end user because nullscope is an artificial construct to make the tests easier to write. As such I believe the correct solution is to change the affected tests to a more realistic cppcheck run which doesn't involve this being null.

Reply:
I've made the change to remove nullScope instead generating a more complete symboldatabase in the tests that used it. Change is in 8860de36523e37310fbc72f6a000d7e64a0123ce . Previously when check was not null and token was null isVariableDeclaration would have crashed rather than returning false as the test expects. This was masked by the different behavior exercised by the test due to check being null. I've added an extra check at the start of the test for tok being null and simply return false now as the test expects. This is in 9b28db290d558fab7a8f03377d21d84c3541114f